Joint nowcasting forecasting (#70)
nfidd/sismid: sessions/forecasting-nowcasting.qmd

```{r libraries, message = FALSE}
library("nfidd")
library("dplyr")
library("tidyr")
library("ggplot2")
library("tidybayes")
library("scoringutils")
```

::: {.callout-tip}
The best way to interact with the material is via the [Visual Editor](https://docs.posit.co/ide/user/ide/guide/documents/visual-editor.html) of RStudio.
:::

## Initialisation

We set a random seed for reproducibility. 
Setting this ensures that you should get exactly the same results on your computer as we do.
We also set an option that makes `cmdstanr` show line numbers when printing model code.
This is not strictly necessary but will help us talk about the models.

```{r}
set.seed(123)
options(cmdstanr_print_line_numbers = TRUE)
```

:::

# The challenge: forecasting with incomplete data

As we've seen in previous sessions, epidemiological data often arrives with delays.
When we want to forecast future trends, we ideally want to use the most recent data available.
However, this recent data is incomplete due to reporting delays.

> The traditional approach is to wait until data is "complete" (or nearly so) before forecasting.

Author: > common practice is to either forecast directly from partially observed data

we cover this one in the original nowcasting session so we can call back to this

>emove recent observations that are assumed to be partially observed

this is the baseline method we are dealing with here. 

>  or make ad hoc adjustments or rescaling to recent partially observed observations.

remove ad-hoc and this becomes a nowcasting method. I will add a callout somewhere about multiplicative "no model" rescalings (i.e. baselinenowcast and friends)

  make_daily_infections(infection_times), gen_time_pmf, ip_pmf
)
head(onset_df)

# Set our analysis cutoff point
cutoff <- 61
forecast_horizon <- 14  # We'll forecast 14 days ahead
```

## Apply reporting delays

Now we'll simulate reporting delays and create our reporting triangle:

```{r simulate-reporting-delays}
reporting_delay_pmf <- censored_delay_pmf(
  rlnorm, max = 15, meanlog = 1, sdlog = 0.5
)
plot(reporting_delay_pmf)
```

```{r simulate-reporting-triangle}
reporting_triangle <- onset_df |>
  filter(day < cutoff + forecast_horizon) |>
  mutate(
    reporting_delay = list(
      tibble(d = 0:15, reporting_delay = reporting_delay_pmf)
    )
  ) |>
  unnest(reporting_delay) |>
  mutate(
    reported_onsets = rpois(n(), onsets * reporting_delay)
  ) |>
  mutate(reported_day = day + d)

# Filter to what would be observed by the cutoff
filtered_reporting_triangle <- reporting_triangle |>
  filter(reported_day <= cutoff)

tail(filtered_reporting_triangle)
```

Now we create the available onsets by summing reported onsets by day - this represents what we observe at the cutoff:

```{r counts-observed}
available_onsets <- filtered_reporting_triangle |>

>   summarise(available_onsets = sum(reported_onsets), .by = day)

Author: its the new syntax gen z are using / hadley finally found the light and copied `data.table`

>   summarise(available_onsets = sum(reported_onsets), .by = day)

Author: I'm not sure I really want people to dwell on this so I am not going to put it in if that is okay. 

TLDR: the difference is this is transitory grouping i.e just in summarise so its the same as group_by followed after the operation by a ungroup.

tail(available_onsets)
```

## Create forecast evaluation datasets

For our analysis, we need two additional datasets:

> 1. **Complete data at forecast horizon**: What we would observe over the forecast horizon

Author: ```suggestion
1. **Complete data at forecast horizon**: What we would observe up to the forecast horizon
```

> 2. **Filtered data for complete data approach**: Data that is "complete" at the time of forecast

Author: ```suggestion
2. **Filtered data for complete data approach**: Data that is "complete" (i.e. expected to be fully reported) at the time of forecast
```

forecast_start_day <- max(complete_data$day)
evaluation_end_day <- max(complete_at_horizon$day)

ggplot() +
  # Final observed data (truth) for evaluation period
  geom_point(data = complete_at_horizon |> 
               filter(day >= forecast_start_day, day <= evaluation_end_day), 
            aes(x = day, y = complete_onsets), 
            color = "black", size = 1) +
  # Complete data forecast ribbons from forecast start
  geom_line(data = complete_data_forecasts |>
              filter(actual_day >= forecast_start_day, actual_day <= evaluation_end_day),
            aes(x = actual_day, y = .value, group = .draw),
            color = "red", alpha = 0.1) +
  # Mark cutoff date (when we actually want forecasts from)
  geom_vline(xintercept = cutoff, 
             linetype = "dashed", color = "blue", linewidth = 1) +
  scale_y_log10() +
  xlim(forecast_start_day, evaluation_end_day) +
  labs(
    x = "Day",
    y = "Onsets",
    title = "Complete data approach",
    subtitle = "Blue line: Forecast date"
  ) +
  theme_minimal()
```

::: {.callout-tip}
## Take 5 minutes

Examine the complete data approach:
- How far back does the forecast have to start from?

> - What happens to forecast uncertainty over the forecasthorizon?

Reviewer: ```suggestion
- What happens to forecast uncertainty over the forecast horizon?
```

```

::: {.callout-tip}
## Take 5 minutes

Compare this pipeline approach to the complete data approach:
- What additional information does it use?
- What are the limitations of using point estimates from the nowcast?
- How does this approach differ from the complete data approach?
- How might model mismatch between the geometric and renewal models affect results?

:::

::: {.callout-note collapse="true"}
## Solution
**What additional information does it use?**
- This approach uses all available data, including the recent truncated observations that were ignored in the complete data approach.

**What are the limitations of using point estimates from the nowcast?**
- It doesn't include the uncertainty from the nowcast, leading to overconfident predictions.
- Point estimates ignore the full posterior distribution from the nowcasting step.

**How does this approach differ from the complete data approach?**
- It has lower variance than the complete data approach due to using more information.
- It appears slightly more biased towards under-predicting than the complete data approach.

**How might model mismatch between the geometric and renewal models affect results?**

> - The geometric random walk and renewal equation represent different generative processes [@lison2024], which can lead to inconsistencies as we saw in the [joint nowcasting session](sessions/joint-nowcasting).

Reviewer 2: ```suggestion
- The geometric random walk and renewal equation represent different generative processes [@lisonGenerativeBayesianModeling2024], which can lead to inconsistencies as we saw in the [joint nowcasting session](sessions/joint-nowcasting).
```

I couldn't find the citation as it was in the .bib file.

            aes(x = day, y = complete_onsets), 
            color = "black", size = 1) +
  # Joint model forecast ribbons from cutoff
  geom_line(data = joint_forecasts |>
              filter(actual_day >= forecast_start_day, actual_day <= forecast_end_day),
            aes(x = actual_day, y = .value, group = .draw),
            color = "purple", alpha = 0.1) +
  # Mark cutoff date
  geom_vline(xintercept = cutoff, 
             linetype = "dashed", color = "blue", linewidth = 1) +
  scale_y_log10() +
  xlim(forecast_start_day, forecast_end_day) +
  labs(
    x = "Day",
    y = "Onsets",
    title = "Joint approach: Forecasting from joint nowcast and forecast",
    subtitle = "Blue line: cutoff (forecast start)"
  ) +
  theme_minimal()
```

::: {.callout-tip}
## Take 5 minutes

What are the advantages of the joint model approach?
- How does it address the model mismatch problem?
- What uncertainties does it properly account for?
- Why might it perform better than the pipeline and complete data approaches?
- Why might it perform worse than the complete data approach?

:::

::: {.callout-note collapse="true"}

> ## Solution

Reviewer 2: The formatting for these answers is a little bit weird when rendered. Like a lot of separate non-bulleted sentences.

  In such cases, older data may provide a more stable basis for forecasting, even though it is less timely.  
  It may also underperform if the generative process is not consistent with the data we have. For example, if the growth rate of older data happens to be more consistent with the future growth rate of the epidemic, then the complete data approach will perform better.
  These issues indicate that the underlying generative process or forecast model could be improved to better capture changes in the epidemic or reporting dynamics.

:::

# Evaluation and comparison

We'll evaluate all four approaches against what will finally be reported (our `complete_at_horizon` dataset) using proper scoring rules.

## Prepare evaluation data

```{r prepare-evaluation}
# Create evaluation dataset - what we're forecasting against
evaluation_data <- complete_at_horizon |>
  filter(day > cutoff, day <= cutoff + forecast_horizon) |>
  mutate(
    forecast_horizon = day - cutoff,
    observed = complete_onsets
  ) |>
  select(day, forecast_horizon, observed)

head(evaluation_data)
```

## Visualise all the forecasts together

As in all other sessions, we first visualise the forecasts together.

> ```{r plot-all-forecasts}

Reviewer 2: I can't really distinguish the overlap/spread of the green/blue lines in particular. Maybe facet? or use sample-derived intervals?

forecast_results <- bind_rows(
  complete_data_forecasts |>
    filter(actual_day > cutoff),
  pipeline_forecasts,
  joint_forecasts
)

forecast_results |>
  ggplot(aes(x = actual_day, y = .value, group = .draw, color = approach)) +
  geom_line(alpha = 0.1) +
  geom_point(data = evaluation_data, aes(x = day, y = observed, group = NULL, color = NULL), color = "black", size = 1) +
  scale_y_log10() +
  labs(x = "Day", y = "Onsets", title = "All forecast approaches") +
  theme_minimal() +
  guides(color = guide_legend(
    title = "Approach", override.aes = list(alpha = 1, linewidth = 1))
  ) +
  theme(legend.position = "bottom")
```

## Quantitative evaluation using scoringutils

Using `scoringutils`, We first create a forecast sample object

```{r forecast-evaluation}
sc_forecasts <- forecast_results |>
    left_join(evaluation_data, by = c("actual_day" = "day")) |>
    group_by(approach, actual_day) |>
    mutate(
      .draw = row_number()
    ) |>
    as_forecast_sample(
      forecast_unit = c("approach", "actual_day"),
      observed = "observed",
      predicted = ".value",
      sample_id = ".draw"
    )
sc_forecasts

> ```

Reviewer 2: ``` 
Warning: Some forecasts have different numbers of rows (e.g. quantiles or samples).
scoringutils found: 98, 80. This may be a problem (it can potentially distort
scores, making it more difficult to compare them), so make sure this is
intended.
```

Does this mean that some models had more forecast horizons than others? if so, maybe consider aligning them so it's the same?

  ) +
  theme_minimal() +
  theme(legend.position = "bottom")
```

::: {.callout-tip}
## Take 5 minutes

Compare the four approaches across the evaluation metrics

:::

::: {.callout-note collapse="true"}
## Solution
From this evaluation, we can see:

- **Complete data approach**: Performs much worse overall, with CRPS of 182 compared to ~19-20 for the other approaches. Performance degrades rapidly with longer forecast horizons.
- **Pipeline approach**: Shows the best overall performance (CRPS = 18.8) but has some systematic underprediction bias (-0.67).
- **Joint approach**: Slightly worse overall performance (CRPS = 20.5) with some overprediction bias (0.69) and higher dispersion, indicating greater uncertainty.

The pipeline approach performs better here potentially because the reporting delay is relatively short and the central nowcast estimate was accurate.
In this case, having less uncertainty was beneficial.
However, this is not guaranteed - the higher dispersion in the joint approach reflects more realistic uncertainty quantification.
If the nowcast had been less accurate or reporting delays longer, the joint approach might have performed better.

:::

> # Going further

Reviewer 2: Somewhere, I'm not sure where, I think it is important to mention the limitation of how these methods as we are presenting them only help when there is systematic under-reporting, correct? Or will they extend to settings where there is adjustments down as well, as happens often with ILI?

Either way, I think this distinction is important to make, given the focus on ILI in the new forecasting content, and talking about how these methods can/should/would need to be adjusted to adapt to that setting will help make the link between the two sets of content be clearer.

Author: > Or will they extend to settings where there is adjustments down as well, as happens often with ILI?

Yes they can be adjusted. We are only dealing with delayed (note not underreporting) as that is the most common bias we observe in real-time linelists (i.e the focus of the first half of the course).

I think somewhere having a map of how a linelist relates to ILI would be useful as this also guides the modelling choices to make.

I don't really want to go into this level of detail here as I think there is enough content to be going on with. Potentially this makes a good talk or call out.

Author: #79

            aes(x = day, y = complete_onsets), 
            color = "black", size = 1) +
  # Joint model forecast ribbons from cutoff
  geom_line(data = joint_forecasts |>
              filter(actual_day >= forecast_start_day, actual_day <= forecast_end_day),
            aes(x = actual_day, y = .value, group = .draw),
            color = "purple", alpha = 0.1) +
  # Mark cutoff date
  geom_vline(xintercept = cutoff, 
             linetype = "dashed", color = "blue", linewidth = 1) +
  scale_y_log10() +
  xlim(forecast_start_day, forecast_end_day) +
  labs(
    x = "Day",
    y = "Onsets",
    title = "Joint approach: Forecasting from joint nowcast and forecast",
    subtitle = "Blue line: cutoff (forecast start)"
  ) +
  theme_minimal()
```

::: {.callout-tip}
## Take 5 minutes

What are the advantages of the joint model approach?
- How does it address the model mismatch problem?
- What uncertainties does it properly account for?
- Why might it perform better than the pipeline and complete data approaches?
- Why might it perform worse than the complete data approach?

:::

::: {.callout-note collapse="true"}
## Solution

- **How does it address the model mismatch problem?**  

>   The joint model uses a single consistent generative process, such as the renewal equation, for both nowcasting and forecasting.  

Author: ```suggestion
  - The joint model uses a single consistent generative process, such as the renewal equation, for both nowcasting and forecasting.  
```
